### PR TITLE
Fix macOS metadata directory cleanup

### PR DESCRIPTION
**Pull request by uczltw6**
## Summary

- remove `__MACOSX` directories as well as `.DS_Store` files before dataset archives are created
- keep symbolic links on the existing unlink path instead of recursively following them
- extend the existing data utility test with real file and directory metadata entries

## Why

`zip_directory()` calls `delete_dsstore()` before creating an archive, and `__MACOSX` is one of the cleanup names enabled by default. The cleanup loop previously called `Path.unlink()` for every match, which raises when the match is a directory and prevents the archive from being created. This change uses `shutil.rmtree()` for real directories while retaining `unlink()` for files and links.

## Tests

- `pytest tests/test_python.py::test_data_utils -q --disable-warnings` — passed
- related download/archive tests — 4 passed
- `ruff format --check ultralytics/utils/downloads.py tests/test_python.py`
- `ruff check --ignore BLE001 ultralytics/utils/downloads.py tests/test_python.py`
- `python -m py_compile ultralytics/utils/downloads.py tests/test_python.py`
- `git diff --check`

AI assistance was used to help identify the edge case and prepare the focused patch. The behavior and test results were verified locally.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Updated dataset archive cleanup to safely remove macOS metadata directories and files, preventing directory unlink errors during archive creation.

### 📊 Key Changes
- Enhanced `delete_dsstore()` to remove real directories with `shutil.rmtree()`.
- Preserved `Path.unlink()` for files and symbolic links to avoid recursively following links.
- Extended `test_data_utils` with `__MACOSX`, `.DS_Store`, and AppleDouble metadata fixtures and cleanup assertions.

### 🎯 Purpose & Impact
Dataset archives can now be created when macOS metadata directories are present, while symbolic links retain the existing unlink behavior.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @uczltw6, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms your contribution was received, and an Ultralytics engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/utils/downloads.py [hunk 1]
```diff
@@ -94,7 +94,10 @@ def delete_dsstore(path: str | Path, files_to_delete: tuple[str, ...] = (".DS_St
         matches = list(Path(path).rglob(file))
         LOGGER.info(f"Deleting {file} files: {matches}")
         for f in matches:
-            f.unlink()
+            if f.is_dir() and not f.is_symlink():
+                shutil.rmtree(f)
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/downloads.py:98]: 💡 **MEDIUM**: If matching directories are nested, such as `root/__MACOSX/nested/__MACOSX`, `matches` contains both paths and the outer directory is processed first. `shutil.rmtree(f)` then removes the nested match, so the later iteration reaches the `else` branch for a nonexistent path and raises `FileNotFoundError`, preventing `zip_directory()` from creating the archive. Process directory matches deepest-first before calling `rmtree()`.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Reviewed the cleanup implementation, archive call path, and focused test changes. The main fix correctly handles real metadata directories and preserves symlink unlink behavior, but nested matching directories can still cause cleanup to fail due to stale paths.

<details><summary>💬 Posted 1 inline comment</summary>

- 💡 **MEDIUM** `ultralytics/utils/downloads.py:98` If matching directories are nested, such as `root/__MACOSX/nested/__MACOSX`, `matches` contains both paths and the outer directory is processed first. `shutil.rmtree(f)` then removes the nested match, so the later iteration reaches the `else` branch for a nonexistent path and raises `FileNotFoundError`, preventing `zip_directory()` from creating the archive. Process directory matches deepest-first before calling `rmtree()`.
</details>

**Comment by uczltw6** [comment]
I have read the CLA Document and I sign the CLA

File: ultralytics/utils/downloads.py (hunk 1 above)
Comment (uczltw6) [reply ultralytics/utils/downloads.py:98]: Fixed in 3e0a68b. Metadata matches are now processed deepest-first, so removing an outer `__MACOSX` directory cannot leave a later stale nested path. The regression creates nested metadata directories and reproduces the previous `FileNotFoundError`; the focused test and related archive tests now pass.

**Review by glenn-jocher — APPROVED** [review]
Cold-reviewed the live head. Recursive deletion is implemented in the existing metadata cleanup owner, handles nested  directories, and the focused  path passes. LGTM.
